server.py

```diff
@@ -1,182 +1,273 @@
-# ADD A PRODUCT TO A SHOPCART
-######################################################################
-@app.route('/shopcarts/<int:uid>/products', methods=['POST'])
-def add_product(uid):
-    """Add a product to the shopcart with the given uid"""
-    cart = Shopcart.find(uid)
-
-    if not cart:
-        return jsonify("Cart with id '{}' was not found.".format(uid)), status.HTTP_404_NOT_FOUND
-
-    try:
-        cart.add_products(request.get_json())
-        cart.save()
-    except DataValidationError as e:
-        message = { 'error': e.args[0] }
-        return jsonify(message), status.HTTP_400_BAD_REQUEST
-
-    return make_response(jsonify(cart.serialize()), status.HTTP_200_OK)
-
-######################################################################
-# DELETE A PRODUCT FROM A SHOPCART
-######################################################################
-@app.route('/shopcarts/<int:uid>/products/<int:pid>', methods=['DELETE'])
-def delete_product(uid, pid):
-    cart = Shopcart.find(uid)
-    if cart:
-        cart.delete_product(pid)
-    return make_response('', status.HTTP_204_NO_CONTENT)
-
-
-
-######################################################################
-#  U T I L I T Y   F U N C T I O N S
-######################################################################
-
-def check_content_type(content_type):
-    """ Checks that the media type is correct """
-    if request.headers['Content-Type'] == content_type:
-        return
-    #app.logger.error('Invalid Content-Type: %s', request.headers['Content-Type'])
-    abort(status.HTTP_415_UNSUPPORTED_MEDIA_TYPE, 'Content-Type must be {}'.format(content_type))
-
-
-
-######################################################################
-# List all Shopcarts
-######################################################################
-@app.route('/shopcarts', methods=['GET'])
-def get_all_shopcarts():
-    pid = request.args.get('pid')
-    if pid:
-        carts = Shopcart.find_by_product( int(pid) )
-    else:
-        carts = Shopcart.all()
-    message = [cart.serialize() for cart in carts]
-    rc = status.HTTP_200_OK
-    return jsonify(message), rc
-
-######################################################################
-# Prune empty Shopcarts
-######################################################################
-@app.route('/shopcarts/prune', methods=['DELETE'])
-def prune_empty_shopcarts():
-    Shopcart.prune()
-    return make_response('', status.HTTP_204_NO_CONTENT)
-
-if __name__ == "__main__":
-    # dummy data for server testing
-    app.run(host='0.0.0.0', port=int(PORT), debug=DEBUG)
+import os
+import json
+from flask import Flask, jsonify, request, url_for, make_response, abort
+from flask_api import status
+from flask_sqlalchemy import SQLAlchemy
+from sqlalchemy import create_engine
+
+from models.shopcart import Shopcart
+from models.dataerror import DataValidationError
+
+# Local database credentials

```

Comment:
You're supposed to get the database credentials in the `models.py` file, not in `server.py`. Homework instructions say the code should mimic the professor's `models.py` file [here](https://github.com/nyu-devops/lab-flask-bdd/blob/43d4d82d015ede6f566e7ef5785deb34758bf36b/app/models.py#L175), but using MySQL instead of Redis. We technically still have our model files split into `product.py` and `shopcart.py`, but we can combine them into a single file.

Reply:
We have to pass db.Model to the Product class for SQLAlchemy. If we put Product class in anther file, product.py has to import server.py. Then there's circular dependencies. My solution is put the Product class in server.py. You guys have better solutions?

Reply:
We're supposed to do all of the database work in `models.py` to ensure separation. So creating the SQLAlchemy db, etc. should be in models.py, and that way you only have `server.py` importing `models.py`. 

We make the database a property of the model, like he did [here](https://github.com/nyu-devops/lab-flask-bdd/blob/43d4d82d015ede6f566e7ef5785deb34758bf36b/app/models.py#L45). If you scroll down, line 174 to the end of the file takes care of reading database credentials/connecting to the database by setting the `.redis` property of the Pet model. He's using Redis, not MySQL, so the code looks different, but it's the same idea.

The corresponding `server.py` file in the repo, you can see all it does is:
`from app.models import Pet`, then uses `Pet.redis` to perform all database calls.

Reply:
Cool, now I know how to do it.